### PR TITLE
feat: update auto-instrumentation to re-inject instrumentation path after init

### PR DESCRIPTION
**Pull request by herin049**
# Description

Updates auto-instrumentation behavior to re-inject instrumentation path after init. This fixes auto-instrumentation with subprocesses (e.g. Uvicorn workers)

Fixes https://github.com/open-telemetry/opentelemetry-python/issues/4927

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

```bash
uv run tox -e py312-test-opentelemetry-instrumentation-wrapt2
```

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -56,6 +56,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#4049](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/4049))
 - `opentelemetry-instrumentation-sqlalchemy`: implement new semantic convention opt-in migration
   ([#4110](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/4110))
+- `opentelemetry-instrumentation`: add ability to auto-instrument subprocesses with `OTEL_PYTHON_AUTO_INSTRUMENTATION_INSTRUMENT_SUBPROCESSES` environment variable
```
Comment (MikeGoldsmith) [inline CHANGELOG.md]: Needs moving up to unreleased.

**Review by MikeGoldsmith — APPROVED** [review]
Looks good - thanks @herin049. Need to move changelog.

**Review by DylanRussell — APPROVED** [review]

**Review by emdneto — APPROVED** [review]
Nice. If merged, I think it's also valid to document this option in opentelemetry.io docs as well + explaining the behavior to users.

File: opentelemetry-instrumentation/README.rst [hunk 2]
```diff
@@ -114,6 +114,14 @@ If set by the user to `patch_all` , opentelemetry instrument will call the geven
 This is considered experimental but can be useful to instrument gevent applications.
 e.g OTEL_PYTHON_AUTO_INSTRUMENTATION_EXPERIMENTAL_GEVENT_PATCH=patch_all
 
+* ``OTEL_PYTHON_AUTO_INSTRUMENTATION_INSTRUMENT_SUBPROCESSES``
+
+By default, auto-instrumentation removes the auto-instrumentation ``sitecustomize.py`` directory
```
Comment (aabmass) [inline opentelemetry-instrumentation/README.rst]: Would be good to state why this is the default. For context https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1050

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py [hunk 3]
```diff
@@ -129,8 +130,17 @@ def initialize(*, swallow_exceptions: bool = True) -> None:
...
+            OTEL_PYTHON_AUTO_INSTRUMENTATION_INSTRUMENT_SUBPROCESSES, ""
+        )
+        .strip()
+        .lower()
+        == "true"
+    )
+
     # prevents auto-instrumentation of subprocesses if code execs another python process
-    if "PYTHONPATH" in environ:
+    if "PYTHONPATH" in environ and not instrument_subprocesses:
```
Comment (aabmass) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py]: For context https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1050

As I was thinking about it, it would probably be safe to re-patch the `PYTHONPATH` after the OTel initialization is finished, when sitecustomize finishes running. Then we could avoid exposing this option at all and make it the default again.

I might be missing some edge cases where the shallow behavior is still preferred and safer. WDYT?

Comment (herin049) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py]: Agreed, I like this approach. Will update the PR.

Comment (aabmass) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py]: Thanks! I think we just need some test coverage

Comment (herin049) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py]: Refactored to use the approach discussed.

File: opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py
Comment (aabmass) [inline opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py:1]: Is it possible to add a test that repros the original issue of subprocess loop, just to be sure the fix is good and no regressions?

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py [hunk 4]
```diff
@@ -157,3 +146,37 @@ def initialize(*, swallow_exceptions: bool = True) -> None:
...
+    # where subprocesses spawned in the initialization phase execute the
+    # initialization phase again, spawning more subprocesses.
+    if python_path is not None:
+        environ["PYTHONPATH"] = _python_path_without_directory(
+            python_path, filedir, pathsep
+        )
+
+    try:
+        _initialize(swallow_exceptions=swallow_exceptions)
+    finally:
```
Comment (aabmass) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:176]: Would it work to just save the original value of PYTHONPATH envvar and restore it here?

**Review by aabmass — APPROVED** [review]

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py (hunk 4 above)
Comment (herin049) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:176]: I wanted to handle the edge case where `PYTHONPATH` is mutated during the initialization phase. Let me know if you think this is overkill, and I can remove this logic.

File: opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py
Comment (herin049) [reply opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py:1]: Good idea, will update the tests.

**Comment by emdneto** [comment]
Thanks for the PR!

Just a heads-up: we no longer update `CHANGELOG.md` directly. The changelog is now generated from changelog fragments using Towncrier.

Please add the appropriate changelog fragment for this change instead of editing `CHANGELOG.md` manually. You can find the instructions and expected format in [CONTRIBUTING.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md#changelog).

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py (hunk 4 above)
Comment (aabmass) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:176]: SG and makes sense.

File: opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py
Comment (aabmass) [reply opentelemetry-instrumentation/tests/auto_instrumentation/test_initialize.py:1]: SG, please move to "ready to merge" once it's done

**Review by tammy-baylis-swi — APPROVED** [review]
Awesome

**Review by xrmx — APPROVED** [review]
pylint is complaining
